### PR TITLE
feat: add trigonometry functions

### PR DESCRIPTION
**Pull request by thisisnic**
This PR adds definitions for common trigonometry functions.

### CODE REVIEW

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [reply extensions/functions_arithmetic.yaml:181]: Adding to this, in Arrow-c++, we only have kernels for fp32 and fp64 (I believe the docs may be wrong here and imply support for decimal inputs but we insert an implicit cast).  You won't get an error if you try the integer types but it will insert an implicit cast.

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [reply extensions/functions_arithmetic.yaml:225]: There is some good input from @jvanstraten on this topic (well, related topic at least) here: https://github.com/substrait-io/substrait/pull/230#issuecomment-1170230630

**Comment by thisisnic** [comment]
Apologies; should have put this in the original PR.  The reasoning for specifying the overflow behaviour options for tangent only was on the basis that the ranges for the returned value for this function are wider than the other ones and so I assumed it'd need it.

My trig is also a bit rusty, but from what I can gather from a bit of a search, in terms of output values:
* range of the cosine and sine functions: [-1, 1]
* range of  tangent: [-infinity, infinity].  
* range of arctangent, arcsine: [-pi/2, pi/2]
* range of arccosine: [0,pi]

Though as @westonpace has said, the approach suggested by @jvanstraten might be more appropriate here.

File: extensions/functions_arithmetic.yaml
Comment (thisisnic) [reply extensions/functions_arithmetic.yaml:181]: I only included all of the others to follow how some other functions have been specified, but I don't see any value in doing it this way if we don't need to.

File: extensions/functions_arithmetic.yaml
Comment (thisisnic) [reply extensions/functions_arithmetic.yaml:225]: Apologies; should have put this in the original PR.  The reasoning for specifying the overflow behaviour options for tangent only was on the basis that the ranges for the returned value for this function are wider than the other ones and so I assumed it'd need it.

My trig is also a bit rusty, but from what I can gather from a bit of a search, in terms of output values:
* range of the cosine and sine functions: [-1, 1]
* range of  tangent: [-infinity, infinity].  
* range of arctangent, arcsine: [-pi/2, pi/2]
* range of arccosine: [0,pi]

Though as @westonpace has said, the approach suggested by @jvanstraten might be more appropriate here.

Comment (jvanstraten) [reply extensions/functions_arithmetic.yaml:225]: `tan` is technically undefined at `(1/2 + k) pi` for all integer `k`, because that's a division by zero when expressed as `sin(x) / cos(x)`. However, due to pi's transcendental nature, none of those numbers can be exactly represented with floating-point numbers, so `tan(x)` is defined for all finite floating-point values of x. So `tan` only needs the rounding option.

`asin` and `acos` do have domain errors; they are undefined for `x < -1` and `x > 1`. So they should be defined however `sqrt` ends up being defined. `atan2` is also a bit weird when `0, 0` is passed, since it's basically `atan(y / x)` (with some boundary conditions to deal with `x = 0` when `y != 0` and to get all quadrants to behave nicely), so I guess it should get the same treatment.

The other trig functions are defined for all finite real numbers, and thus only need rounding options.

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [reply extensions/functions_arithmetic.yaml:181]: I think, if the end result is no different than a cast the kernel is not needed.  If there are other cases like this we should get rid of those kernels as well.

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:256]:
```suggestion
    description: "Get the arccosine of a value in radians."
```

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:278]:
```suggestion
    description: "Get the arcsine of a value in radians."
```

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:300]:
```suggestion
    description: "Get the arctangent of a value in radians."
```

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [inline extensions/functions_arithmetic.yaml:322]:
```suggestion
    description: "Get the arctangent of values given as x/y pairs."
```
The inputs to atan2 are not actually in radians

**Review by westonpace — COMMENTED** [review]
Sorry, minor nits

I don't know about the hyphen.  [Merriam-webster](https://www.merriam-webster.com/dictionary/arccosine) gives one word without hyphen.  [Wolfram Alpha](https://reference.wolfram.com/language/ref/ArcCos.html) and [cpp reference](https://en.cppreference.com/w/cpp/numeric/math/acos) give two words without hyphen.

File: extensions/functions_arithmetic.yaml
Comment (westonpace) [reply extensions/functions_arithmetic.yaml:225]: > due to pi's transcendental nature

Especially when topped with ice cream

File: extensions/functions_arithmetic.yaml
Comment (ianmcook) [reply extensions/functions_arithmetic.yaml:181]: Perhaps we should allow an impl here to specify multiple allowed types for a single value arg, e.g.
```yaml
    impls:
      - args:
           - value: [ i8, i16, i32, i64, fp32,  fp64 ]
         return: fp64
```
That would reduce the amount of boilerplate in these YAML files and at the same time avoid Substrait producers needing to insert lots of fussy explicit casts.

Comment (jacques-n) [reply extensions/functions_arithmetic.yaml:181]: @ianmcook we explored patterns around this [early](https://github.com/substrait-io/substrait/commit/9e84da55393a24953ed9f9869fa423f86f4860f7) but it became a slippery slope of complexity for implementers. Ultimately, we figured that since yamls aren't something that will be worked with super-frequently it was ok to be a bit more verbose to simplify implementation of consumption of the yaml.

Comment (ianmcook) [reply extensions/functions_arithmetic.yaml:181]: >Adding to this, in Arrow-c++, we only have kernels for fp32 and fp64 (I believe the docs may be wrong here and imply support for decimal inputs but we insert an implicit cast). You won't get an error if you try the integer types but it will insert an implicit cast.

@westonpace Is that correct? In an Acero ExecPlan, I believe `add_checked(int8, int8)` returns `int8`

Comment (ianmcook) [reply extensions/functions_arithmetic.yaml:181]: @jacques-n that makes sense, but this seems to create a dilemma in which we must decide to do one or the other of these things:

1. Have producers include many explicit casts in Substrait plans (even when they might not strictly be necessary) which I believe adds computational cost for engines executing these plans
2. Add impls to these YAML files covering every combination of input types and the associated output type, which creates a minor combinatorial explosion even just for the binary functions

Comment (jacques-n) [reply extensions/functions_arithmetic.yaml:181]: A few thoughts:

* I don't understand why the explicit casts would cost extra compute if they are already being done implicitly. 
* I agree with Weston's comment: "If there are other cases like this we should get rid of those kernels as well." 
* From the Substrait perspective: 2 is the "right" approach. Define the specific things that are possible. Avoid trying to be overly fancy in the function definition paradigm. Use a script to define things if you want. (This isn't a place we need optimization, generally.)

Comment (ianmcook) [reply extensions/functions_arithmetic.yaml:181]: To clarify what I mean in the comment above: consider the case of `add(i32, i64)`. In every engine of which I am aware, that returns `i64`. If you reverse the order of the value args so it's `add(i64, i32)`, that also returns `i64`. So should we add impls for `add` like so?
```yaml
      - args:
          - options: [ SILENT, SATURATE, ERROR ]
            required: false
          - value: i32
          - value: i64
        return: i64
      - args:
          - options: [ SILENT, SATURATE, ERROR ]
            required: false
          - value: i64
          - value: i32
        return: i64
```
and repeat this for every permutation of input for which engines tend to behave the same way? I read through https://github.com/substrait-io/substrait/issues/65 and it all seems reasonable but does not speak to this question.

Comment (jacques-n) [reply extensions/functions_arithmetic.yaml:181]: Yes, the current pattern requires this. 

Do you have an i32 + i64? Is it because of some performance benefit? I would expect that to be rare. The most common pattern systems have is an implementation of i64 + i64 and they cast i32 to i64 before doing i64 + i64.

Comment (thisisnic) [reply extensions/functions_arithmetic.yaml]: I've updated to just support i64 and fp64 now.

Comment (westonpace) [reply extensions/functions_arithmetic.yaml]: I think `fp32` is probably more valuable than `i64` and I apologize for being obnoxious :).  My goal isn't to limit the number of kernels but to create one kernel per distinct implementation.  I think it is common for hardware/engines to have different implementations for float and double.  By "different implementation" I mean "passing the same inputs yields a different result"

For an example, consider the [C++ standard library implementation for cosine](https://en.cppreference.com/w/cpp/numeric/math/cos).  There are three ways to call `std::cos`, with a `float`, with a `double`, or with an `integral` type.  However, calling cosine with an integral type is identical to casting to double and calling cosine with that double value.  We can see this both in the docs:

> 4) A set of overloads or a function template accepting an argument of any [integral type](https://en.cppreference.com/w/cpp/types/is_integral). Equivalent to 2) (the argument is cast to double).

and in practice:

https://godbolt.org/z/h6Pb5WY86

Comment (thisisnic) [reply extensions/functions_arithmetic.yaml]: Thanks for going into detail about that, makes sense!  Updated to now have `fp32 ` and `fp64`

File: extensions/functions_arithmetic.yaml
Comment (thisisnic) [reply extensions/functions_arithmetic.yaml]: As the conversation regarding those other options now appears to be resolved, have updated the options here in line with those suggestions.

**Review by jvanstraten — DISMISSED** [review]
LGTM

File: extensions/functions_arithmetic.yaml
Comment (jacques-n) [inline extensions/functions_arithmetic.yaml:198]: Same question as elsewhere, are these options in the order of most preferred to least preferred? Remember that an engine that support multiple option values on a non-specified value will pick them in the order they are declared.

Comment (jvanstraten) [reply extensions/functions_arithmetic.yaml:198]: `TIE_TO_EVEN` a.k.a. "round to nearest" is [the default mode used by glibc](https://www.gnu.org/software/libc/manual/html_node/Rounding.html), and as such is probably what everyone is already implicitly doing, so I placed that first intentionally.

Side note; reading that description I guess my earlier assumptions that "even" means "even number" is wrong, and it actually refers to the least significant bit of the mantissa. I don't think that changes anything for us though.

File: extensions/functions_arithmetic.yaml
Comment (jvanstraten) [reply extensions/functions_arithmetic.yaml]: Part of this discussion is somewhat related to https://github.com/substrait-io/substrait/issues/251, so linking for posterity.

**Comment by jacques-n** [comment]
This looks good to me. Looks like it needs a conflict resolved before merge, however. 

@jvanstraten , are you happy with this?

**Comment by jvanstraten** [comment]
Yep, LGTM

**Comment by jacques-n** [comment]
@thisisnic, can you do a conflict resolution?

**Comment by thisisnic** [comment]
@jvanstraten Apologies, for some reason when I fixed the conflict and force-pushed, it dismissed your review.

**Review by jvanstraten — APPROVED** [review]
Still LGTM :smile:

**Comment by jacques-n** [comment]
Thanks @thisisnic !

**Pull request merged**
